### PR TITLE
Rename fields

### PR DESCRIPTION
**Pull request by vladvelici**
Rename Message fields and fix tests for history. Enable materialised history tests.

## Release Notes

- **Documentation**
  - Updated the README.md to clarify message handling, including real-time updates and versioning.
  
- **New Features**
  - Enhanced message handling with a focus on versioning for updates and deletions.
  - Introduced new properties in message objects, such as `version`, `operation`, and improved error handling.

- **Bug Fixes**
  - Improved tests to ensure accurate message handling, including checks for new properties and behaviors.

- **Tests**
  - Expanded integration and unit tests to validate new message structures and behaviors, ensuring robustness in message processing.

### CODE REVIEW

**Comment by vladvelici** [comment]
@coderabbitai full review

File: src/core/messages.ts [hunk 1]
```diff
@@ -496,9 +496,10 @@ export class DefaultMessages
       text,
       metadata ?? {},
       headers ?? {},
-      new Date(response.createdAt),
       ChatMessageActions.MessageCreate,
       response.serial,
+      new Date(response.createdAt),
```
Comment (splindsay-92) [inline src/core/messages.ts:501]: Maybe worth just adding a small comment to say, timestamp and createdAt are equal, or doing above const timestamp = response.createdAt so its explicit that this logic was wanted?

File: src/core/messages.ts [hunk 2]
```diff
@@ -510,50 +511,54 @@ export class DefaultMessages
...
     const deletedMessage: Message = new DefaultMessage(
       message.serial,
       message.clientId,
-      message.roomId,
+      this._roomId,
       message.text,
-      message.metadata,
-      message.headers,
-      message.createdAt,
+      (message.metadata as typeof message.metadata | undefined) ?? {},
```
Comment (splindsay-92) [inline src/core/messages.ts:549]: I'm curious, but why is this now needed here?

**Review by splindsay-92 — COMMENTED** [review]
Submitting what I have so far just to help speed things along, but I'm still reviewing :)

File: src/core/message.ts [hunk 3]
```diff
@@ -207,62 +214,65 @@ export class DefaultMessage implements Message {
...
-  actionEqual(message: Message): boolean {
+  versionEqual(message: Message): boolean {
     // Check to ensure the messages are the same before comparing operation order
     if (!this.equal(message)) {
-      throw new ErrorInfo('actionEqual(): Cannot compare actions, message serials must be equal', 50000, 500);
+      throw new ErrorInfo('versionEqual(): Cannot compare versions, message serials must be equal', 50000, 500);
     }
 
-    return this.latestActionSerial === message.latestActionSerial;
+    return this.version === message.version;
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/message.ts:275]: Version is a string that can safely be compared with <, > and ===.

File: test/core/chat.integration.test.ts [hunk 4]
```diff
@@ -60,6 +60,7 @@ describe('Chat', () => {
     expect(occupancy).toEqual(expect.objectContaining({ connections: 1, presenceMembers: 0 }));
 
     // Request history, and expect it to succeed
+    await new Promise((resolve) => setTimeout(resolve, 3000)); // wait for cassandra
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply test/core/chat.integration.test.ts:63]: good suggestion generally but these awaits are temporary so I'll keep it as it is for now

File: src/core/messages.ts [hunk 5]
```diff
@@ -496,9 +496,10 @@ export class DefaultMessages
       text,
       metadata ?? {},
       headers ?? {},
-      new Date(response.createdAt),
       ChatMessageActions.MessageCreate,
       response.serial,
+      new Date(response.createdAt),
+      new Date(response.createdAt),
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/messages.ts:502]: It's intentional. One date is for createdAt, one is for timestamp. They should point to the same value, and that value is represented by `createdAt` after a message was posted.

File: src/core/index.ts [hunk 6]
```diff
@@ -29,10 +29,11 @@ export type { LogContext, Logger, LogHandler } from './logger.js';
 export { LogLevel } from './logger.js';
 export type {
   Message,
-  MessageActionDetails,
+  Operation as MessageActionDetails,
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/index.ts:32]: oh this is weird

File: test/core/messages.integration.test.ts [hunk 7]
```diff
@@ -255,24 +267,32 @@ describe('messages integration', () => {
...
+        timestamp: deletedMessage1.deletedAt,
+        action: ChatMessageActions.MessageDelete,
       }),
     ]);
 
+    // test shorthand getters
+    expect(history.items[0]?.isUpdated).toBe(false);
+    expect(history.items[0]?.isDeleted).toBe(true);
+    expect(history.items[0]?.deletedAt).toEqual(deletedMessage1.deletedAt);
+    // expect(history.items[0]?.deletedBy).toEqual(deletedMessage1.deletedBy);
```
Comment (splindsay-92) [inline test/core/messages.integration.test.ts:288]: Should this be un-commented?

File: test/core/messages.integration.test.ts [hunk 8]
```diff
@@ -255,24 +267,32 @@ describe('messages integration', () => {
...
+
     // We shouldn't have a "next" link in the response
     expect(history.hasNext()).toBe(false);
   });
 
   // At the moment, the history API does not materialize updated messages in the history.
```
Comment (splindsay-92) [inline test/core/messages.integration.test.ts:294]: This is no longer the case, can we remove the comment?

File: test/core/messages.integration.test.ts [hunk 9]
```diff
@@ -243,7 +255,7 @@ describe('messages integration', () => {
   });
 
   // At the moment, the history API does not materialize deleted messages in the history.
```
Comment (splindsay-92) [inline test/core/messages.integration.test.ts:257]: Same here, the comment isn't accurate now

File: test/core/messages.integration.test.ts [hunk 10]
```diff
@@ -288,20 +308,26 @@ describe('messages integration', () => {
...
+        timestamp: updatedMessage1.updatedAt,
+        action: ChatMessageActions.MessageUpdate,
       }),
     ]);
 
+    // test shorthand getters
+    expect(history.items[0]?.isUpdated).toBe(true);
+    expect(history.items[0]?.isDeleted).toBe(false);
+    expect(history.items[0]?.updatedAt).toEqual(updatedMessage1.updatedAt);
+    // expect(history.items[0]?.updatedBy).toEqual(updatedMessage1.updatedBy);
```
Comment (splindsay-92) [inline test/core/messages.integration.test.ts:329]: Same here, should this be commented out or not? Can we uncomment or remove it please?

File: src/core/message-parser.ts [hunk 11]
```diff
@@ -62,48 +49,40 @@ export function parseMessage(roomId: string | undefined, inboundMessage: Ably.In
...
 
   switch (message.action) {
     case ChatMessageActions.MessageCreate: {
       break;
     }
     case ChatMessageActions.MessageUpdate:
     case ChatMessageActions.MessageDelete: {
-      if (!message.version) {
-        throw new Ably.ErrorInfo(`received incoming ${message.action} without version`, 50000, 500);
+      if (!message.createdAt) {
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/message-parser.ts:62]: Seems to be always set, will simplify this. Thanks for noticing!

File: README.md [hunk 12]
```diff
@@ -304,17 +304,17 @@ const updatedMessage = await room.messages.update(message,
...
+In rare occasions updates might arrive over realtime out of order. To keep a correct state, compare the `version` lexicographically (string compare). Alternatively, the `Message` interface provides convenience methods to compare two instances of the same base message to determine which version is newer: `versionBefore()`, `versionAfter()`, and `versionEqual()`.
 
-The same out-of-order situation can happen between updates received over realtime and HTTP responses. In the situation where two concurrent edits happen, both might be received via realtime before the HTTP response of the first one arrives. Always use `actionAfter()`, `actionBefore()`, or `actionEqual()` to determine which instance of a `Message` is newer.
+The same out-of-order situation can happen between updates received over realtime and HTTP responses. In the situation where two concurrent edits happen, both might be received via realtime before the HTTP response of the first one arrives. Always compare the message `version` to determine which instance of a `Message` is newer.
```
Comment (splindsay-92) [inline README.md:317]: small nitpick *edit* -> *update*

File: src/core/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/index.ts:32]: sorted

File: src/core/message.ts [hunk 13]
```diff
@@ -99,71 +99,78 @@ export interface Message {
...
 
   /**
-   * The details of the latest action that updated the message. This is only set for update and delete actions.
+   * The timestamp at which this version was updated, deleted, or created.
    */
-  readonly latestActionDetails?: MessageActionDetails;
+  readonly timestamp: Date;
+
+  /**
+   * The details of the latest operation that updated the message. This is only set for update and delete actions.
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/message.ts:115]: done

File: src/core/message-parser.ts [hunk 14]
```diff
@@ -62,48 +49,40 @@ export function parseMessage(roomId: string | undefined, inboundMessage: Ably.In
...
     }
     default: {
       throw new Ably.ErrorInfo(`received incoming message with unhandled action; ${message.action}`, 50000, 500);
     }
   }
+
+  const timestamp = message.timestamp;
+  const createdAt = message.createdAt ?? timestamp;
+
   return new DefaultMessage(
```
[the remark this answers is not in the document]
Comment (vladvelici) [reply src/core/message-parser.ts:75]: Yeah, I think that would be good. Either change the constructor or have a static method that takes an object and calls the constructor correctly.

File: test/core/messages.integration.test.ts (hunk 7 above)
Comment (vladvelici) [reply test/core/messages.integration.test.ts:288]: I'm adding a PR to fix this in realtime first (nearly done), I don't think we should wait for this PR though

File: src/core/messages.ts (hunk 1 above)
Comment (vladvelici) [reply src/core/messages.ts:501]: added comment

File: test/core/messages.integration.test.ts (hunk 10 above)
Comment (vladvelici) [reply test/core/messages.integration.test.ts:329]: added comment to both of these, a todo to uncomment when operation is returned from the endpoint

File: src/core/message-parser.ts [hunk 15]
```diff
@@ -62,48 +49,32 @@ export function parseMessage(roomId: string | undefined, inboundMessage: Ably.In
...
-    newMessage.operation,
+    message.serial,
+    message.clientId,
+    roomId,
+    message.data.text,
+    message.data.metadata ?? {},
+    message.extras.headers ?? {},
+    message.action as ChatMessageActions,
+    message.version,
+    new Date(message.createdAt),
```
Comment (splindsay-92) [inline src/core/message-parser.ts:76]: Do we still need a check above to ensure createdAt and timestamp are present?

Comment (vladvelici) [reply src/core/message-parser.ts:76]: We might do a cleanup of this function in the future

**Review by splindsay-92 — APPROVED** [review]

File: src/core/message-parser.ts (hunk 15 above)
Comment (vladvelici) [reply src/core/message-parser.ts:76]: Turns out we don't need the MessagePayload interface anymore so I've removed that

**Pull request merged**